### ISSUE TITLE
ArcGIS_InstallPatch failed to execute Test-TargetResource functionality while trying to install patch

### ISSUE THREAD

**Issue by NabinPaudel** [Esri/arcgis-powershell-dsc#423]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Issue
PowerShell DSC resource ArcGIS_InstallPatch  failed to execute Test-TargetResource functionality with error message: Patch File G:\Software\ArcGIS10.9.1\Patches\server\ArcGIS-1091-S-K1-Patch.msp **is not accessible**
At C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.0\ArcGIS.psm1:261 char:5

At C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.0\ArcGIS.psm1:261 char:5

### Module Version
Modules\ArcGIS\4.0.0

### Affected Resource(s)

* ArcGIS_Server
* ArcGIS_Portal
* ArcGIS_DataStore

### Configuration Files

```
{
    "AllNodes": [
        {
            "NodeName": "gis-s",
            "DataStoreTypes": [
                "Relational"
            ],
            "Role": [
                "Server",
                "Portal",
                "DataStore",
                "ServerWebAdaptor",
                "PortalWebAdaptor"
            ]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",
	"OldVersion": "10.8.1",
        "ServerContext": "server",
        "PortalContext": "portal",
        "ServerRole": "GeneralPurposeServer",
        "DownloadSetups": false,
        "DownloadPatches": true,
        "Credentials": {
            "ServiceAccount": {
                "UserName": "gis",
                "Password": "*************",
                "IsDomainAccount": true,
                "IsMSAAccount": false
            }
        },
        "Server": {
            "LicenseFilePath": "G:\\Software\\ArcGIS10.9.1\\ArcGISGISServerAdvanced_ArcGISServer_1128037.prvc",
            "Installer": {
                "Path": "G:\\Software\\ArcGIS10.9.1\\ArcGIS_Server_Windows_1091_180041.exe",
                "InstallDir": "F:\\ArcGIS\\Server",
                "InstallDirPython": "C:\\Python27",
                "PatchesDir": "G:\\Software\\ArcGIS10.9.1\\Patches\\server",
                "PatchInstallOrder": [
                    
                ],
                "EnableArcMapRuntime": false,
                "EnableDotnetSupport": false
            },
            "ServerDirectoriesRootLocation": "F:\\arcgisserver\\directories",
            "ConfigStoreLocation": "F:\\arcgisserver\\config-store",
            "LocalRepositoryPath": "F:\\arcgisserver\\local",
            "ServerLogsLocation": "F:\\arcgisserver\\logs",
            "ServerDirectories": [
                {
                    "name": "arcgiscache",
                    "physicalPath": "F:\\arcgisserver\\directories\\arcgiscache",
                    "directoryType": "CACHE",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                },
                {
                    "name": "arcgisjobs",
                    "physicalPath": "F:\\arcgisserver\\directories\\arcgisjobs",
                    "directoryType": "JOBS",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 360
                },
                {
                    "name": "arcgisoutput",
                    "physicalPath": "F:\\arcgisserver\\directories\\arcgisoutput",
                    "directoryType": "OUTPUT",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 10
                },
                {
                    "name": "arcgissystem",
                    "physicalPath": "F:\\arcgisserver\\directories\\arcgissystem",
                    "directoryType": "SYSTEM",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }
            ],
            "RegisteredDirectories": [
                {
                    "name": "BasemapsCacheDir",
                    "physicalPath": "F:\\arcgisserver\\additionaldirs\\cache",
                    "directoryType": "CACHE"
                },
                {
                    "name": "TestOutputDir",
                    "physicalPath": "F:\\arcgisserver\\additionaldirs\\testOutput",
                    "directoryType": "OUTPUT"
                },
                {
                    "name": "TestJobsDir",
                    "physicalPath": "F:\\arcgisserver\\additionaldirs\\testJobs",
                    "directoryType": "JOBS"
                }
            ],
            "PrimarySiteAdmin": {
                "UserName": "admin",
                "Password": "********"
            }
        },
        "Portal": {
            "LicenseFilePath": "G:\\Software\\ArcGIS10.9.1\\ArcGIS_Enterprise_Portal_1091_371728_20220113.json",
            "PortalLicenseUserTypeId": "creatorUT",
            "Installer": {
                "Path": "G:\\Software\\ArcGIS10.9.1\\Portal_for_ArcGIS_Windows_1091_180052.exe",
                "WebStylesPath": "",
                "InstallDir": "F:\\ArcGIS\\Portal",
                "ContentDir": "F:\\arcgisportal",
                "PatchesDir": "G:\\Software\\ArcGIS10.9.1\\Patches\\portal",
                "PatchInstallOrder": [
                    
                ]
            },
            "ContentDirectoryLocation":"F:\\arcgisportal\\content",
            "EnableAutomaticAccountCreation": true,
            "DefaultRoleForUser": "",
            "DefaultUserLicenseTypeIdForUser": "viewerUT",
            "PortalAdministrator": {
                "UserName": "admin",
                "Email": "",
                "Password": "",
                "SecurityQuestionIndex": ,
                "SecurityAnswer": ""
            },
            "EmailSettings":{
                "SMTPServerAddress":"",
                "From":"",
                "Label":"SBX No Reply",
                "AuthenticationRequired":false,
                "SMTPPort":"",
                "PasswordFilePath": "",
                "EncryptionMethod": "NONE"
            }
        },
        "DataStore": {
            "ContentDirectoryLocation": "F:\\arcgisdatastore",
            "EnableFailoverOnPrimaryStop": false,
            "EnablePointInTimeRecovery": false,
            "Installer": {
                "Path": "G:\\Software\\ArcGIS10.9.1\\ArcGIS_DataStore_Windows_1091_180054.exe",
                "InstallDir": "F:\\ArcGIS\\DataStore",
                "InstallAllFeatures": false,
                "PatchesDir": "G:\\Software\\ArcGIS10.9.1\\Patches\\datastore",
                "PatchInstallOrder": [
                    
                ]
            }
        },
        "WebAdaptor": {
            "AdminAccessEnabled": true,
            "Installer": {
                "Path": "G:\\Software\\ArcGIS10.9.1\\ArcGIS_Web_Adaptor_for_Microsoft_IIS_1091_180055.exe",
                "PatchesDir": "",
                "PatchInstallOrder": [
                    
                ]
            },
            "WebSiteId": 1
        }
    }
}
```

### Expected Behavior
The patch should have installed from the folder specified. Note, we wanted to control which patch are installed by downloading only patch that we want installed in the patch folder path which is supplied in json.

### Actual Behavior

```
Trace-DSCJob : 10/3/2022 1:44:40 PM: PowerShell DSC resource ArcGIS_InstallPatch  failed to execute Test-TargetResource functionality with error message: Patch File G:\Software\ArcGIS10.9.1\Patches\server\ArcGIS-1091-S-K1-Patch.msp is not accessible
At C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.0\ArcGIS.psm1:261 char:5
+     Trace-DSCJob -Job $Job -JobName $ConfigurationName -DebugMode $De ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Trace-DSCJob

Trace-DSCJob : 10/3/2022 1:44:40 PM: PowerShell DSC resource ArcGIS_InstallPatch  failed to execute Test-TargetResource functionality with error message: Patch File G:\Software\ArcGIS10.9.1\Patches\portal\ArcGIS-1091-PFA-Log4j-PatchB.msp is not accessible
At C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.0\ArcGIS.psm1:261 char:5
+     Trace-DSCJob -Job $Job -JobName $ConfigurationName -DebugMode $De ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Trace-DSCJob

Trace-DSCJob : 10/3/2022 1:44:40 PM: PowerShell DSC resource ArcGIS_InstallPatch  failed to execute Test-TargetResource functionality with error message: Patch File G:\Software\ArcGIS10.9.1\Patches\datastore\ArcGIS-1091-DS-DE-Patch.msp is not accessible
At C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.0\ArcGIS.psm1:261 char:5
+     Trace-DSCJob -Job $Job -JobName $ConfigurationName -DebugMode $De ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Trace-DSCJob
```

### Steps to Reproduce
```
 Invoke-ArcGISConfiguration -ConfigurationParametersFile C:\Temp\single-machine-base-deployment-s-fs-1091-upgrade-patches.json -Mode Install
```
### Important Factoids

The installation was done prior to patching using PoweShell DSC.

**Comment by cameronkroeker** [comment]
Hi @NabinPaudel,

Couple questions:

1. Did DSC create the PatchesDir locations, or were these location manually created prior? 
- `G:\\Software\\ArcGIS10.9.1\\Patches\\datastore`
- `G:\\Software\\ArcGIS10.9.1\\Patches\\portal`
- `G:\\Software\\ArcGIS10.9.1\\Patches\\server`

When `"DownloadPatches": true,`, DSC will create those locations if they do not exist which should set the correct permissions and be accessible. However, if they were manually created before the DSC run perhaps the permissions are not set.

Also, if they were manually created, is `G:\` a local drive or is it a mapped network drive?

Thanks,
Cameron K.

**Comment by NabinPaudel** [comment]
Hi @cameronkroeker,
The patches folders were created manually prior and patches were downloaded prior to install in those folders.

I missed to mention in the issue description that, when "DownloadPatches": true, It downloads the complete set of patches and installs them fine. However when _"DownloadPatches":false_ i get the mentioned error.

We want to control only specific patches are installed by manually downloading required patches. But if we use _"DownloadPatches": true_, it downloads & overwrites the folder content with all the patches and install them all. Could you advise if there is any other way of controlling only specific patches are installed? Or setting up the correct permission to the folder where patches are downloaded prior.

**Comment by cameronkroeker** [comment]
> Hi @cameronkroeker, The patches folders were created manually prior and patches were downloaded prior to install in those folders.
> 
> I missed to mention in the issue description that, when _"DownloadPatches": true_, It downloads the complete set of patches and installs them fine. However when _"DownloadPatches":false_ i get the mentioned error.
> 
> We want to control only specific patches are installed by manually downloading required patches. But if we use _"DownloadPatches": true_, it downloads & overwrites the folder content with all the patches and install them all. Could you advise if there is any other way of controlling only specific patches are installed? Or setting up the correct permission to the folder where patches are downloaded prior.

Hi @NabinPaudel,

In that case you will want to perform the following:

1. Ensure **ConfigData.DownloadPatches** attribute is set to false (this is the default value).
2. Define an existing path for **ConfigData.Server.Installer.PatchesDir** attribute, and ensure the patches are already downloaded and exist in the specified location.
3. If you don't define array of patch .msp file names in the **ConfigData.Server.Installer.PatchInstallIOrder** attribute, then all the patches in **ConfigData.Server.Installer.PatchesDir** are installed in a random order.

> Note: When ConfigData.DownloadPatches is true you can control which patches are downloaded and installed, by defining the patches in ConfigData.Server.Installer.PatchInstallIOrder attribute. If there aren't any patches defined in ConfigData.Server.Installer.PatchInstallIOrder then it will download and install ALL available patches for the version, The install order is based release date of the patch.

Thanks,
Cameron K.

**Comment by NabinPaudel** [comment]
@cameronkroeker resolved with workaround mentioned in #411 for the scenario of 
```
"DownloadPatches": false
```
